File: web/src/features/specs/impact.ts
/**
 * Pure model of the requirement impact view (story GIT-US-0131, doc 03
 * §21.11): the URL state of the base/head pickers, the tier grouping of the
 * hits, the per-tier status line, the reason codes read aloud and the ref
 * suggestions the pickers offer.
 *
 * Tier 3 candidates are a group of their own and never sit among the certain
 * hits of tiers 1 and 2 (R-IMP-4): a candidate is a guess with a score, not a
 * reached requirement.
 */

import type { SearchSchemaInput } from '@tanstack/react-router';
import { z } from 'zod';

import type { ImpactResult, SyncRepoStatus } from '@/api/provider';

/** One requirement a diff affects (doc 03 §21.11 R-IMP-5). */
export type ImpactHit = ImpactResult['hits'][number];

/** The head value that names the working tree; the query sends no `head` for it. */
export const WORKTREE = 'worktree';

/** The base the view diffs against when the URL names none: the usual default branch. */
export const DEFAULT_BASE = 'main';

export type ImpactSearch = { base?: string; head?: string };

const refParam = z
  .preprocess(
    (value) => (typeof value === 'string' ? value.trim() : undefined),
    z.string().max(200).optional(),
  )
  .catch(undefined);

const impactSearchSchema = z.object({ base: refParam, head: refParam });

/** Parses raw search params. Never throws; an empty value is dropped. */
export function parseImpactSearch(input: unknown): ImpactSearch {
  const result = impactSearchSchema.safeParse(input ?? {});
  if (!result.success) return {};
  const { base, head } = result.data;
  return {
    ...(base ? { base } : {}),
    ...(head ? { head } : {}),
  };
}

/** Route `validateSearch` for `/p/$project/specs/impact`. */
export function validateImpactSearch(input: ImpactSearch & SearchSchemaInput): ImpactSearch {
  return parseImpactSearch(input);
}

/** True when a head value means the working tree: empty, or the `worktree` keyword. */
export function isWorktree(head: string | undefined): boolean {
  return head === undefined || head === '' || head.toLowerCase() === WORKTREE;
}

/** The range the view asks for: `base` defaulted, `head` omitted for the working tree. */
export function impactRange(search: ImpactSearch): { base: string; head?: string } {
  const base = search.base ?? DEFAULT_BASE;
  return isWorktree(search.head) ? { base } : { base, head: search.head as string };
}

/** `main..worktree`, the way the report's text form names a range. */
export function formatRange(base: string, head: string | undefined): string {
  return `${base}..${isWorktree(head) ? WORKTREE : head}`;
}

export type TierKey = 1 | 2 | 3;

export type TierStatus = ImpactResult['tiers'][number];

export type TierGroup = {
  tier: TierKey;
  /** The tier's own status line; absent when the answer did not report the tier. */
  status?: TierStatus;
  hits: ImpactHit[];
};

export const tierLabels: Record<TierKey, { title: string; hint: string }> = {
  1: { title: 'Direct', hint: 'The diff touches a traced file, symbol or marker.' },
  2: { title: 'Transitive', hint: 'A changed symbol is called from traced code.' },
  3: {
    title: 'Candidates',
    hint: 'Semantically close to the change; not reached by a trace or a call.',
  },
};

/** The tier a hit is shown under: a candidate is always tier 3, whatever else it says. */
export function tierOf(hit: ImpactHit): TierKey {
  return hit.candidate ? 3 : hit.tier;
}

/**
 * The hits grouped by tier, always three groups in tier order. Tiers 1 and 2
 * keep the answer's order (spec, then number, R-IMP-6); candidates are ranked
 * by score, highest first, then by ref (R-IMP-8).
 */
export function groupByTier(result: Pick<ImpactResult, 'tiers' | 'hits'>): TierGroup[] {
  return ([1, 2, 3] as const).map((tier) => {
    const hits = result.hits.filter((hit) => tierOf(hit) === tier);
    if (tier === 3) {
      hits.sort((a, b) => (b.score ?? 0) - (a.score ?? 0) || a.ref.localeCompare(b.ref));
    }
    const status = result.tiers.find((t) => t.tier === tier);
    return { tier, hits, ...(status ? { status } : {}) };
  });
}

/** The sentence of a tier's status; `undefined` for a plain `ok`. */
export function tierStatusText(status: TierStatus | undefined): string | undefined {
  if (!status) return 'Not reported.';
  switch (status.status) {
    case 'ok':
      return status.truncated ? 'Partial: the per-symbol limit cut some callers.' : undefined;
    case 'skipped':
      return 'Skipped: this tier was not asked for.';
    case 'unavailable':
      return `Unavailable${status.message ? `: ${status.message}` : '.'} Tiers 2 and 3 need Pando.`;
    case 'error':
      return `Error${status.message ? `: ${status.message}` : '.'}`;
  }
}

export type ReadReason = { kind: string; text: string; title: string };

const reasonKinds: Record<string, string> = {
  file: 'file',
  symbol: 'symbol',
  marker: 'marker',
  renamed: 'renamed',
  removed: 'removed',
  implements: 'implements',
  modifies: 'modifies',
  delta: 'pending delta',
  call: 'called from',
  semantic: 'semantic',
};

const callPattern = /^call:(\S+) calls (\S+) d(\d+)$/;

/**
 * A reason code read aloud (doc 03 §21.11 R-IMP-2..5). `call:<caller> calls
 * <symbol> d<n>` becomes "called from <caller> → <symbol>, depth n"; `+<n>`
 * counts the reasons the answer left out.
 */
export function readReason(reason: string): ReadReason {
  const call = callPattern.exec(reason);
  if (call) {
    const [, caller = '', symbol = '', depth = ''] = call;
    return { kind: 'called from', text: `${caller} → ${symbol}, depth ${depth}`, title: reason };
  }
  if (/^\+\d+$/.test(reason)) {
    return { kind: 'more', text: `${reason.slice(1)} more`, title: reason };
  }
  const colon = reason.indexOf(':');
  if (colon < 0) return { kind: reasonKinds[reason] ?? reason, text: '', title: reason };
  const key = reason.slice(0, colon);
  return { kind: reasonKinds[key] ?? key, text: reason.slice(colon + 1), title: reason };
}

/** The ref suggestions of the base picker, then the head picker. */
export type RefSuggestions = { base: string[]; head: string[] };

/**
 * What the pickers offer: the current branch and its upstream from the sync
 * status, the default branch, `HEAD` and its recent ancestors, and — for the
 * head — the working tree first.
 */
export function refSuggestions(rows: readonly SyncRepoStatus[] | undefined): RefSuggestions {
  const branches: string[] = [];
  for (const row of rows ?? []) {
    const status = row.status;
    if (!status) continue;
    if (!status.detached && status.branch && status.branch !== '@') branches.push(status.branch);
    if (status.upstream) branches.push(status.upstream);
  }
  const recent = ['HEAD', 'HEAD~1', 'HEAD~2', 'HEAD~5'];
  const unique = (values: string[]) => [...new Set(values.filter(Boolean))];
  return {
    base: unique([DEFAULT_BASE, ...branches, ...recent]),
    head: unique([WORKTREE, ...branches, ...recent]),
  };
}

File: internal/core/impact.go
package core

import "errors"

// The answer types of the requirement impact query (ADR-037, docs/03 section
// 21.11, GIT-US-0119): given a diff, the requirements it affects, each with
// the tier that reached it, the reasons, its coverage state and whether it is
// suspect. The query is resolved by a native host — it needs git, the marker
// scan and, for tiers 2 and 3, Pando — so only the shapes a host hands back
// through the vault seam live here, free of any filesystem access. Nothing
// here is ever written into a file: an impact report is derived data.

// ErrUnknownRevision is a base or head an impact query names that the
// repository's history does not hold.
var ErrUnknownRevision = errors.New("unknown revision")

// The three impact tiers, cheapest and most certain first.
const (
	// ImpactTierDirect: a changed file or symbol carries a marker or a
	// trace: entry of the requirement, or the story the diff is for names it
	// in its Spec Delta or its links.
	ImpactTierDirect = 1
	// ImpactTierTransitive: Pando's call graph reaches, from a changed
	// symbol, a caller that carries a marker of the requirement.
	ImpactTierTransitive = 2
	// ImpactTierSemantic: Pando's semantic search over requirement blocks
	// ranks the requirement near the changed symbol names and the story
	// title. A candidate, never a certainty.
	ImpactTierSemantic = 3
)

// ImpactTierStatus says how one tier of an impact query was answered.
type ImpactTierStatus string

// The tier states.
const (
	// ImpactTierOK: the tier ran; its hits are in the result.
	ImpactTierOK ImpactTierStatus = "ok"
	// ImpactTierUnavailable: the tier needs a backend this session has not
	// got, or that did not answer (no Pando, Pando unreachable or timing out).
	ImpactTierUnavailable ImpactTierStatus = "unavailable"
	// ImpactTierError: the backend answered with an error of its own, for
	// example a code project Pando has not indexed.
	ImpactTierError ImpactTierStatus = "error"
	// ImpactTierSkipped: the caller did not ask for the tier.
	ImpactTierSkipped ImpactTierStatus = "skipped"
)

// ImpactQuery is what an impact query resolves.
type ImpactQuery struct {
	// Base is the revision the diff starts from: a branch, a SHA, HEAD.
	Base string `json:"base"`
	// Head is the revision the diff ends at; empty means the working tree.
	Head string `json:"head,omitempty"`
	// Story is the story or task the diff is for. Its Spec Delta and its
	// implements / modifies links are direct hits, and its title feeds the
	// semantic tier.
	Story ItemID `json:"story,omitempty"`
	// Title is extra text for the semantic tier, next to the story title.
	Title string `json:"title,omitempty"`
	// Tiers limits the query to these tiers; empty runs all three.
	Tiers []int `json:"tiers,omitempty"`
	// Depth is the transitive caller depth of tier 2; 0 is the host default.
	Depth int `json:"depth,omitempty"`
	// Limit caps the callers per changed symbol of tier 2 and the candidates
	// of tier 3; 0 is the host default.
	Limit int `json:"limit,omitempty"`
}

// Wants reports whether the query asks for a tier.
func (q ImpactQuery) Wants(tier int) bool {
	if len(q.Tiers) == 0 {
		return true
	}
	for _, t := range q.Tiers {
		if t == tier {
			return true
		}
	}
	return false
}

// ImpactTier is how one tier was answered.
type ImpactTier struct {
	Tier   int              `json:"tier"`
	Status ImpactTierStatus `json:"status"`
	// Hits counts the result's hits whose strongest tier is this one.
	Hits int `json:"hits"`
	// Truncated reports that the tier's input or its backend's answer was cut
	// at a bound, so the tier may reach more than it reports.
	Truncated bool `json:"truncated,omitempty"`
	// Message says why a tier is unavailable or failed.
	Message string `json:"message,omitempty"`
}

// ImpactKind says what a certain (tier 1 or 2) hit's diff changed: the code
// behind the requirement, or only a test that verifies it (GIT-US-0157).
type ImpactKind string

// The hit kinds.
const (
	// ImpactKindBehaviour: at least one reason reaches the requirement
	// through its code — an Implements: marker or a trace.code entry, directly
	// or through a call — or the story names it in its links or Spec Delta.
	ImpactKindBehaviour ImpactKind = "behaviour"
	// ImpactKindTestOnly: every reason reaches the requirement through a test
	// that verifies it — a Verifies: marker or a trace.tests entry. The tests
	// changed; what the requirement states may not have.
	ImpactKindTestOnly ImpactKind = "test-only"
)

// ImpactHit is one requirement a diff affects. A requirement reached by
// several tiers is one hit, at its strongest (lowest) tier, listing the
// reasons of every tier that reached it with certainty.
type ImpactHit struct {
	Ref   RequirementRef `json:"ref"`
	Title string         `json:"title"`
	Tier  int            `json:"tier"`
	// Kind is behaviour or test-only on a tier-1 or tier-2 hit; empty on a
	// tier-3 candidate, which is not a trace.
	Kind ImpactKind `json:"kind,omitempty"`
	// Candidate marks a tier-3 hit: a semantic neighbor, not a trace.
	Candidate bool `json:"candidate,omitempty"`
	// Score is the semantic score of a candidate, rounded to three decimals.
	Score float64 `json:"score,omitempty"`
	// Status is the requirement's coverage state (docs/03 section 21.6);
	// empty when no coverage backend answered.
	Status CoverageStatus `json:"status,omitempty"`
	// Suspect is set when the coverage state is suspect, or when passing
	// evidence covers code the diff changes directly or transitively.
	Suspect bool `json:"suspect,omitempty"`
	// Reasons are short codes, "<kind>:<detail>" (docs/03 section 21.11),
	// sorted by tier and then by text, at most a few then "+<n>".
	Reasons []string `json:"reasons"`
	// Pending lists the open stories and tasks whose unapplied Spec Delta
	// modifies the requirement (R-DELTA-10).
	Pending []ItemID `json:"pending,omitempty"`
}

// ImpactResult is the answer of an impact query.
type ImpactResult struct {
	Base string `json:"base"`
	Head string `json:"head,omitempty"`
	// Files counts the changed files; Symbols the changed symbols they map to.
	Files   int          `json:"files"`
	Symbols int          `json:"symbols"`
	Tiers   []ImpactTier `json:"tiers"`
	Hits    []ImpactHit  `json:"hits"`
}
